### PR TITLE
Add service for single point of interaction with database

### LINKED ISSUE #82: Application does not terminate if database connection fails
**Describe the bug**
- since various components of the application depend on a mongodb database, it **should terminate** if the connection is not made.

**To Reproduce**
1. kill your MongoDB instance
2. try running the system

**Expected behavior**
1. since MongoDB is a hard dependency of the project, it must terminate if connection is not made.

**Screenshots**
[asciicast](https://asciinema.org/a/401523)

**Desktop (please complete the following information):**
 - OS: Ubuntu 19.04

### LINKED ISSUE #88: Centralize Database operations
## Current design
all components of the application (handlers) interact with the database separately. The problems associated with this:

1. Not reusing the connection.
2. Frequent connect-disconnect cycles to the database.
3. **Redundant codebase** and associated issues
    - all calls need to establish a connection.
    - repeated access to environment variables and need to assign default cases. Even Minor change requires edit in multiple files that essentially do the same thing.
    - set configurations (`{useUnifiedTopology:true, ...}`) in multiple places. Difficult to ship updates or fix bugs as all calls have separate behavior if configurations do not match.

**Describe the solution you'd like**
- a central provider that contains and exports all database operations that the entire project uses.
- this would fix the problems mentioned in the current implementation.

**Describe alternatives you've considered**
- we could rely on some ORM like mongoose, but considering that we're using only basic operations in the database, using an ORM like mongoose will slow down the application. 

**Additional context**
- This will also provide a method to implement the requested features in #82

### LINKED ISSUE #89: Rewrite database logic to avoid callbacks
**Is your feature request related to a problem? Please describe.**
- the current implementation uses callbacks extensively for database operations.
- this makes the code non-linear and difficult to improve/debug.
- code becomes redundant as each layer of callback requires its own handling of errors.
- often referred to as callback-hell 🔥 

**Describe the solution you'd like**
- Since the database driver supports promises, we can either use the promise syntax or the async-await flow as an alternative. This will keep the codebase linear flow.

This will also allow us to document the codebase better as it is practically easier to document small functions than nested arrangements of them.

**Describe alternatives you've considered**
- we can use third-party libraries like [async.js](https://caolan.github.io/async/v3/) but I personally not support adding libraries for functionality that can be easily done by built-in functions. 

**Additional context**
- a detailed resource for [callback hell](http://callbackhell.com/)

### PR DESCRIPTION
**Pull request by YashKumarVerma**
## Summary
There are three issues linked with the project, here's a summary of what this PR accomplishes.

- a central point of database interaction, provides a thin abstraction over mongdb driver. Takes care of points raised in #88.
- A tiny utility function to kill the process if database connections have failed after multiple tries. As discussed in #82 
- Rewrite of functions using the async-await syntax (replacing callbacks, fallbacks explained in #89 ) 

## Details
- util.js for commonly used utility functions that are consumed by database services, like data transformers.
- exhaustive JSDocs for each member function, with due references to MongoDB documentation.
- All functions which were previously scattered across the application (and also duplicated) combined into one class, eliminating redundancy in code.
- Single point to change the behavior of database connections
- single point connector that provides connection instances to other modules
- kills the process if the connection repeatedly fails.
- all the places which were initially interacting with the database have been refactored to use to database service along with the async-await syntax.
 

## Motivation
- Fixes #82 
- Fixes #88 
- Fixes #89 

## Testing
- service tested by running along with the system, unit tests are in process of being written
- Screencast showing termination of process with NZEC as the connection was not successful.
[asciicast](https://asciinema.org/a/402535)

## Questions
- Suggestions are welcomed.
- please suggest a suitable name for the PR.

### CODE REVIEW

**Comment by birm** [comment]
Have you done any other configuration changes compared to the defaults in develop.yml? I can't get it to read from mongo at all.

**Comment by YashKumarVerma** [comment]
just checked, the problem was that the connector function was appending `mongodb://` in the passed connection from process.env, fixing it

**Comment by YashKumarVerma** [comment]
Hello there, is there something that I can improve in the PR ??

**Comment by birm** [comment]
I'm still not getting any results from any of the find endpoints despite data being in the database. Similarly, posting slide metadata doesn't update the mongo collections or the backend's slide/find route's results.
If you've done the usual test (add slide, view slide, delete slide) successfully, then perhaps there's something more deeply different than expected about our environments or something of the sort.

**Comment by YashKumarVerma** [comment]
Changelog:
- add unit tests for components of database service
- log errors of db service components
- handle cases of invalid objectID

**Review by birm — CHANGES_REQUESTED** [review]
I'm still entirely unable to get this to work within the camicroscope development deployment.
What issues are you having with Distro?

**Comment by YashKumarVerma** [comment]
So I tried debugging the issue, isolated the service, and tested it, but couldn't find the area of the problem.

Then I did the following, based on routes.json, created a postman collection of the endpoints that operate on database actions. 

image

The endpoints were behaving as expected, but there was one patten: the listing route was not working.

**The problem was a simple typo in dataHandler that was calling mongo.add in place of mongo.find**

The tests I wrote were testing the service, but the handler, therefore it was not detected in the tests.

**Comment by YashKumarVerma** [comment]
I've deployed my fork at:http://yashkumarverma-workbench.uaenorth.cloudapp.azure.com:4010/apps/table.html

**Comment by YashKumarVerma** [comment]
@birm ping :)

**Review by birm — COMMENTED** [review]
It seems to be working better than before, I'll just have to make sure nothing else was broken.

File: service/database/index.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+          x["_id"] = {
+            $oid: x["_id"],
+          };
+        });
+      }
+
+      return data;
+    } catch (e) {
+      console.error(e);
+      return e;
```
Comment (birm) [inline service/database/index.js:41]: Does this error return trigger setting the response code appropriately?

Comment (YashKumarVerma) [reply service/database/index.js:41]: thanks for pointing this !

I isolated two functions to test the behavior, looks like the return statement has to be replaced with a throw statement. Below is a simple gist to demonstrate the behavior.

https://gist.github.com/232fa53a05d96fb24a925e35ac0e3db2

pushed commits for this change.

**Review by birm — APPROVED** [review]
Thank you for taking the time to go through these edits.

**Comment by YashKumarVerma** [comment]
Thanks! meanwhile got to learn a lot about the project  ⭐

**Pull request merged**